### ISSUE TITLE
[Request] Store input field ID along with login and password

### ISSUE THREAD

**Issue by talha131** [bitwarden/browser#519]
There are number of reported issues that complain about bitwarden auto-filling incorrect fields. For example,

1. https://github.com/bitwarden/browser/issues/442
1. https://github.com/bitwarden/browser/issues/317

Recently I ran into an issue due to this. Bitwarden has my router's admin credential. When I logged in, bitwarden autofilled all the configuration fields like WAN password, Wifi SSID, Wifi password. I had to disable bitwarden for router configuration page as a work around.

You commented [here](https://github.com/bitwarden/browser/issues/442#issuecomment-353096416),

> This would be pretty difficult to distinguish. Right now we assume any password field is what you want to autofill.

What if when storing password and login for a page, bitwarden also stores the input field id? This way bitwarden will never get confused about the fields.

A similar work around is already documented [here](https://github.com/bitwarden/browser/issues/317#issuecomment-356509131),

>  in case you didn't know, there is a solution/workaround...use the Custom Fields feature to add a field called "login:AccessCard" and put the username or card number there. Then leave the "username" field in the vault entry BLANK. Works fine.

As it must be obvious, it is difficult and cumbersome to pick field id from browser's element inspector and make entries manually. But bitwarden can do this job automatically when it saves login for a new page. 

This will also fix following issue, where bitwarden auto fills new account form unnecessarily along with login form,

![](https://i.imgur.com/lZc13OO.png)

To summarize, every username and password entry must have an optional input id field. When saving new logins, bitwarden should also save field id from which it read username and password.

**Comment by kspearrin** [comment]
This would be a perfect solution if the website never changed. The problem is that websites change all the time. So the next time Facebook changes their login form, it's very likely ids for those fields also change.

**Comment by talha131** [comment]
> The problem is that websites change all the time. So the next time Facebook changes their login form, it's very likely ids for those fields also change.

Sure. But in that case bitwarden can fallback to the current algorithm. What I propose is

1. Save input id with username and password
1. When attempting auto-fill, find and fill input fields with matching ids
1. **If matching ids are not found** then
    1. [assume any password field is what you want to autofill](https://github.com/bitwarden/browser/issues/442#issuecomment-353096416)
    1. When user clicks login button, offer to update entry in bitwarden, which will update input ids. This is what bitwarden already does, it offers to save logins. The only difference is this time it is also updating ids.

Let me know what you think.
